Change esp32 to build on top of the tflite-micro-esp-examples idf component

mocleiri opened the pull request:
This is work in progress to change how tensorflow is built and linked into the microlite firmware.  Instead of enumerating all of the source files I build the tensorflow part with esp_nn for accelerated ops directly as an esp-idf component.

It also upgrades the microlite module to work with latest micropython master (>1.20.0).

The submodules are deleted. Instead a new scripts/build.sh has been created.  It has two functions:
1. **prepare_dependencies** - this will create a dependencies subdirectory and setup a shallow clone of: ulab, micropython, tflite-micro-esp-examples, esp32-camera, micropython-camera-driver, tflite-micro
2. **build_esp32** - this will build the esp32 boards. At the moment it works for the boards/esp32/build/MICROLITE board.

I've run through the steps manually and was able to build firmware for the MICROLITE board and flash it an run the microspeech demo.

The second commit contains changes from @uraich for upgrading microlite to use the new module registration macros.

The next steps are to change the ci builds to use the new build script, upgrade the other board configurations and ensure the unix and rp2 builds still work properly.

Commit:
Remove git submodules in favour of shallow checkout approach

Instead of using submodules create a script that will make a shallow git clone of the dependency repos:
1. ulab
2. micropython
3. tflite-micro-esp-examples
4. tflite-micro (for non esp32 boards)

Pin the versions to the latest releases of each dependency.

ulab: 6.0.12
micropython: master
esp32-tflite: v1.0.0
tflite-micro: <undefined>

tflite-micro doesn't tag releases so its undefined to be the latest commit on the tip of the main branch.

use latest micropython to get upgrade to idf 5.0.2

Commit:
[microlite] update for post micropython 1.20.0 module changes

Commit:
[esp32] update MICROLITE board config for latest micropython

I copied the micropython board ESP32_GENERIC and replaced MICROLITE with minor adjustments
to enable the microlite module and the name of the board.

Commit:
[esp32] Update MICROLITE_C3 board configuration

Also update ci for building esp32 MICROLITE using the new build scripts.

Commit:
Fix unix port

bonner232 commented:
Jo how can I download the firmware, it says that its expired.

cnadler86 commented:
Perhaps you might want also to update to the new [camera API for Micropython](https://github.com/cnadler86/micropython-camera-API). This works with the actual Micropython version.

mocleiri committed:
Pin micropython to v1.20.0

Change branch build logic.

mocleiri committed:
Upgrade upload0artifact plugin to v4

mocleiri committed:
Pin micropython to v1.21.0

mocleiri committed:
Add back all esp32 boards

This should trigger all esp32 boards to build.

mocleiri committed:
Adjust esp32 boards build steps

mocleiri committed:
Adjust firmware archive id's to be unique

mocleiri committed:
[esp32] update board configs from upstream

How boards are configured has been simplified in micropython.

Update the board config files from upstream.

mocleiri committed:
[esp32] increase factory partition for S3.
